### PR TITLE
Update for Couchbase 8.0 release

### PR DESCRIPTION
**Pull request by VirajAgarwal-couchbase**
## Description

This PR introduces the updates needed for the Couchbase 8.0 version. Updating the integration to support all of the new features.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ✅ ] New feature (non-breaking change that adds functionality)
- [ ✅ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ✅ ] Documentation update
- [ ✅ ] Code refactoring
- [ ✅ ] Performance improvement
- [ ✅ ] Test update

## Checklist

- [ ✅ ] I have made corresponding changes to the documentation (if applicable)
- [ ✅ ] I have added tests that prove my fix is effective or that my feature works

## Summary by CodeRabbit

* **New Features**
  * Added two Couchbase stores (QueryStore & SearchStore), deprecating the legacy vector store; richer index management, per-field indexing, and expanded metadata filter capabilities.
  * New Couchbase RAG prompts for both Search and Query stores; prompts exported for use in agents.

* **Documentation**
  * Reworked Couchbase docs and README with dual-store guidance, index-type selection, filtering examples, RAG retrieval examples, and compatibility notes.

* **Tests**
  * Extensive unit and integration tests covering filters, translators, index lifecycle, upserts, queries, and edge cases.

* **Chores**
  * Released Couchbase store package as 1.0.0 and updated test Couchbase image to 8.0.0.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: stores/couchbase/src/vector/filters/operators/data_types/date.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+function lt_for_dates(field: string, value: Date): any {
+  if (null_or_undefined_or_nan(value)) {
+    return {};
+  }
+  return {
+    field: field,
+    end: value,
+    inclusive_start: false,
+  };
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/date.ts:54]: This is Couchbase specific. The Couchbase FTS filters require this. Refer here -> https://docs.couchbase.com/cloud/search/search-request-params.html#numeric-range-queries

File: stores/couchbase/src/vector/filters/operators/data_types/date.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  if (null_or_undefined_or_nan(value)) {
+    return {};
+  }
+  return {
+    field: field,
+    end: value,
+    inclusive_end: true,
+    inclusive_start: false,
+  };
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/date.ts:66]: This is Couchbase specific. The Couchbase FTS filters require this. Refer here -> https://docs.couchbase.com/cloud/search/search-request-params.html#numeric-range-queries

File: stores/couchbase/src/vector/filters/operators/data_types/number.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  if (null_or_undefined_or_nan_or_infinite(value)) {
+    return {};
+  }
+  value = normalize_number(value);
+  return {
+    field: field,
+    max: value,
+    inclusive_min: false,
+  };
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/number.ts:65]: This is Couchbase specific. The Couchbase FTS filters require this. Refer here -> https://docs.couchbase.com/cloud/search/search-request-params.html#numeric-range-queries

File: stores/couchbase/src/vector/filters/operators/data_types/number.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    return {};
+  }
+  value = normalize_number(value);
+  return {
+    field: field,
+    max: value,
+    inclusive_max: true,
+    inclusive_min: false,
+  };
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/number.ts:78]: This is Couchbase specific. The Couchbase FTS filters require this. Refer here -> https://docs.couchbase.com/cloud/search/search-request-params.html#numeric-range-queries

File: stores/couchbase/src/vector/filters/operators/data_types/number.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  if (null_or_undefined_or_nan_or_infinite(value)) {
+    return '';
+  }
+  value = normalize_number(value);
+  return `(${field} = ${value})`;
+}
+
+function qv_ne_for_numbers(field: string, value: number): any {
+  return `(${field} != ${value})`;
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/number.ts:129]: Have checked with SQL++ as well, there might be an edge case here regarding trying to mention not equal to Nan for a numbers field. have added the suggestion though.

File: docs/src/content/en/docs/rag/retrieval.mdx [hunk 6]
```diff
@@ -485,6 +485,26 @@ export const ragAgent = new Agent({
...
+  Process queries using the provided context. Structure responses to be concise and relevant.
+  {/* Choose the prompt that matches your vector store */}
+  {/* Note: CouchbaseQueryStore requires at least one vector to be upserted before index creation. */}
+  ${COUCHBASE_SEARCH_STORE_PROMPT}
+  ${COUCHBASE_QUERY_STORE_PROMPT}
+  `,
+  tools: { vectorQueryTool },
+});
+```
+  </TabItem>
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply docs/src/content/en/docs/rag/retrieval.mdx:506]: Have made the update.

File: stores/couchbase/src/vector/filters/operators/data_types/string.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+function qv_eq_for_strings(field: string, value: string): any {
+  if (value === null || value === undefined || value.length === 0) {
+    return '';
+  }
+  return `(${field} = '${value}')`;
+}
+
+function qv_ne_for_strings(field: string, value: string): any {
+  return `(${field} != '${value}')`;
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/string.ts:61]: Ohh, nice catch. Have made the update.

[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/data_types/string.ts:61]: Have made the update.

File: stores/couchbase/src/vector/filters/operators/queryVector_logical_operator.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
+function qv_and(nodes: any[]): any {
+  const filteredNodes = nodes.filter(node => node && node.length !== 0);
+  if (filteredNodes.length === 0) {
+    return '';
+  }
+  if (filteredNodes.length === 1) {
+    return filteredNodes[0];
+  }
+  return `(${filteredNodes.join(' AND ')})`;
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/filters/operators/queryVector_logical_operator.ts:10]: This validation is not needed because the way that these "or", "nor" and "and" translation functions are used is this ->

The logical handler function gets called for one of these operator (at this point the op is just a key in the json), then the logical handler decided how the children of the operator must be handled and returns 2 things. The first function is to handle this level of logic (like "and", "or", "nor" and "not" function) and the 2nd function is to handle the children of this operator key. Then the highest level of function call just calls the lowerLevelHandler first and then pipes the results of those to the ThisLevelHandler. 

Now, even if the ThisLevelHandler (which is one of the "and", "nor", "not" or "or" functions) receives translation of its children as empty objects, it will return an empty object. And if any array based functions higher in the function stack get these empty objects they will just discard them.

Also, since this particular function is part of the functions which translates json to string, the input ideally will not be anything other than strings.

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+            password,
+            bucketName,
+            scopeName,
+            collectionName,
+          },
+        },
+        error,
+      );
+    }
+  }
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts:98]: Have removed it. Thanks CodeRabbit.

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+      results.rows.forEach((row: any) => {
+        if (row.name === indexName) {
+          index = row;
+        }
+      });
+
+      // Extract the index statistics
+      const dimensions = index.with.dimension;
+      const count = -1; // Not added support yet for adding a count of documents covered by an index
+      const metric = index.with.similarity.toUpperCase() as CouchbaseQueryVectorIndexMetric;
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts]: Have added error handling here.

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+
+      // Execute the SQL++ query
+      await this.scope.query(sqlpp_query);
+    } catch (error: any) {
+      // Check for 'already exists' error (Couchbase may throw a 400 or 409, or have a message)
+      const message = error?.message || error?.toString();
+      if (message && message.toLowerCase().includes('index exists')) {
+        await this.validateExistingIndex(indexName, dimension, metric as MastraMetric);
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts:301]: This method is part of the core mastra package. Its definition can be found in the MastraVector (which extends the MastraBase) class.

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+      });
+      this.clusterPromise = baseClusterPromise;
+      this.cluster = null as unknown as Cluster;
+      this.bucketName = bucketName;
+      this.collectionName = collectionName;
+      this.scopeName = scopeName;
+      this.collection = null as unknown as Collection;
+      this.bucket = null as unknown as Bucket;
+      this.scope = null as unknown as Scope;
+      this.vector_dimension = null as unknown as number;
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts]: Have made the changes.

File: stores/couchbase/src/vector/index.integration.test.ts [hunk 13]
```diff
@@ -235,16 +244,29 @@ describe('Integration Testing CouchbaseVector', async () => {
...
     ];
     let testVectorIds: string[] = ['test_id_1', 'test_id_2', 'test_id_3'];
 
     beforeAll(async () => {
+      await couchbase_client.createIndex({
+        indexName: test_indexName,
+        dimension,
+        metric: 'euclidean',
+        fields_to_index: [{ name: 'brightness', type: 'number' }],
+      });
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/index.integration.test.ts:269]: Have made the changes

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+      const results = await this.cluster.query(sqlpp_query);
+      results.rows.forEach((row: any) => {
+        if (row.name === indexName) {
+          index = row;
+        }
+      });
+
+      // Extract the index statistics
+      const dimensions = index.with.dimension;
+      const count = -1; // Not added support yet for adding a count of documents covered by an index
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts:165]: The count is always a number, having it null creates more confusion right?

[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts:165]: have added the note.

File: stores/couchbase/src/vector/queryVectorStore.ts [hunk 15]
```diff
@@ -0,0 +1,589 @@
...
+      await this.getCollection();
+      if (!(await this.listIndexes()).includes(indexName)) {
+        throw new Error(`Index ${indexName} does not exist`);
+      }
+      await this.cluster?.queryIndexes().dropIndex(this.bucketName, indexName, {
+        scopeName: this.scopeName,
+        collectionName: this.collectionName,
+        ignoreIfNotExists: true,
+      });
+      this.vector_dimension = null as unknown as number;
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.ts]: Have made the update.

File: stores/couchbase/src/vector/queryVectorStore.integration.test.ts [hunk 16]
```diff
@@ -0,0 +1,1408 @@
...
+      { stdio: 'inherit' },
+    );
+  } catch (error: any) {
+    console.error('Error creating bucket:', error.message);
+    // Decide if you want to re-throw or handle specific errors here
+  }
+
+  // Wait for cluster to be fully available after potential operations
+  await new Promise(resolve => setTimeout(resolve, 10000));
+}
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.integration.test.ts:53]: Its part of the integration tests. Also, throwing error is not valid here since, one could want to run the tests on an already intialised Couchbase cluster.

File: stores/couchbase/src/vector/queryVectorStore.integration.test.ts [hunk 17]
```diff
@@ -0,0 +1,1408 @@
...
+        const result = await collection.get(idToBeUpdated);
+        expect(result.content.embedding).toEqual(newVector);
+        expect(result.content.metadata).toEqual(newMetaData);
+      } finally {
+        // Cleanup
+        for (const id of ids) {
+          await collection.remove(id);
+        }
+      }
+    });
```
[the remark this answers is not in the document]
Comment (VirajAgarwal-couchbase) [reply stores/couchbase/src/vector/queryVectorStore.integration.test.ts:568]: Its fine. This is also to test how a user might possibly use the function. You are correct it is not used inside the function definition.

**Comment by DanielSLew** [comment]
Hi @VirajAgarwal-couchbase — I've rebased your 3 commits onto the latest `main` and pushed them to `pr-10685-rebased` on this repo: https://github.com/mastra-ai/mastra/tree/pr-10685-rebased

I tried to push directly to your branch but `maintainerCanModify` is disabled on this PR, and the head repo is `couchbaselabs/mastra` (an org repo), so I don't have push access.

**Could you either:**
1. Enable "Allow edits from maintainers" on this PR so we can push updates, or
2. Pull the rebased branch into your `new-vector-indexes-introduction` branch:
   ```
   git fetch https://github.com/mastra-ai/mastra.git pr-10685-rebased
   git reset --hard FETCH_HEAD
   git push --force-with-lease
   ```

The build (`pnpm build --filter=@mastra/couchbase`) passes cleanly on the rebased branch.

**Comment by abhiaiyer91** [comment]
closing due to inactivity
